usaspending_api/common/helpers/endpoint_documentation.py

```diff
@@ -0,0 +1,118 @@
+    a tuple).
+
+    Returns [("url", <RegexURLPattern object>), ...]
+    """
+    from usaspending_api import urls
+
+    results = []
+
+    def _traverse_urls(base, url_patterns):
+        for url in url_patterns:
+            cleaned = base + url.regex.pattern.lstrip("^").rstrip("$")
+            if hasattr(url, "url_patterns"):
+                _traverse_urls(cleaned, url.url_patterns)
+            elif not endpoint_prefixes or cleaned.startswith(endpoint_prefixes):
+                results.append((cleaned, url))
+
+    _traverse_urls("/", urls.urlpatterns)
+
+    return results
+
+
+def get_endpoints_from_endpoints_markdown():
+    """
+    Looks for and extracts URLs for patterns like |[display](url)|method|description|
+    from the master endpoints.md markdown file.
+    """
+    with open(os.path.join(BASE_DIR, ENDPOINTS_MD)) as f:
+        contents = f.read()
+    return [e.split("?")[0] for e in ENDPOINT_PATTERN.findall(contents) if e]
+
+
+def get_fully_qualified_name(obj):
+    """
+    Fully qualifies an object name.  For example, would return
+    "usaspending_api.common.helpers.endpoint_documentation.get_fully_qualified_name"
+    for this function.
+    """
+    return '{}.{}'.format(obj.__module__, obj.__qualname__)
+
+
+def validate_docs(url, url_object, master_endpoint_list):
+    """
+    Ensures that an endpoint_doc property and a docstring is provided for the
+    view associated with the provided url and checks that the URL is mentioned
+    in the master endpoints.md doc.
+    """
+    qualified_name = url_object.lookup_str
+    view_class = url_object.callback.cls
+
+    messages = []
+
+    if not hasattr(view_class, "endpoint_doc"):
+        messages.append("{} ({}) missing endpoint_doc property".format(qualified_name, url))
+    else:
+        endpoint_doc = getattr(view_class, "endpoint_doc")
+        if not endpoint_doc:
+            messages.append("{}.endpoint_doc ({}) is invalid".format(qualified_name, url))
+        else:
+            absolute_endpoint_doc = os.path.join(BASE_DIR, endpoint_doc)
+            if not case_sensitive_file_exists(absolute_endpoint_doc):
+                messages.append(
+                    "{}.endpoint_doc ({}) references a file that does not exist".format(qualified_name, url)
+                )
+
+    if not (view_class.__doc__ or "").strip():
+        messages.append("{} ({}) has no docstring".format(qualified_name, url))
+
+    # Fix up the url a little so we can pattern match it.
+    pattern = url + ("" if url.endswith("/") else "/") + "?"  # Optional trailing slash.
+
+    # See if we can find anything in the master endpoint list that matches this pattern.
+    for endpoint in master_endpoint_list:
+        if re.fullmatch(pattern, endpoint):
+            break
+    else:
+        messages.append("No URL found in {} that matches {} ({})".format(ENDPOINTS_MD, url, qualified_name))
+
+    return messages
+
+
+__all__ = [
```

Comment:
What determined the reason to keep two of the items out of __all__ ? (If I'm interpreting what is the feature of explicitly setting `__all__`

Reply:
They weren't being used anywhere outside of the module.  Honestly, I don't know why I even bothered with the `__all__`.